### DISCUSSION TITLE
Allow default values in row headings to be translatable as content

### DISCUSSION THREAD

**Discussion by foxtrotcz** [craftcms/cms#13703] [Ideas] [answered]
Hello,
specific example is Table field with Row headings as described here: https://github.com/craftcms/cms/pull/13231
If you create such Table you cannot translate the predefined Row headings because they are defined in Field itself as Default values.
So the headings Width, Height and Length from example will stay in English even in other language site.

Right now it can be partially solved by Static message translations which can translate the strings. https://craftcms.com/docs/4.x/sites.html#static-message-translations
This works for Twig output. But it will stay untranslated in control panel which can lead to confusion.
Edit: After the improvement it now gets translated, but according to CP language, not site language. This means if you prefer English CP and edit German site you will see the row headings in English which is wrong since they are part of the content.

If the Field default values could be made translatable for each site/language it would make things easier.

It makes sense if Default values in each field match the target language.

Edit: Specifically for Default values, Row Headings and other strings used in frontend the translation should be according to site language, not CP language. So if you use CP in English and edit German site then Default values and Row Headings should be translated to German.

Thanks.

**Comment by piotrpog** [comment]
Definitely something that should be implemented. We can switch language of control panel and all its built in text content, like "save", "delete" and such, but we cannot do the same for the fields. This is detrimental to multi-lingual teams working on the same project.

Not to mention that some field labels or settings are sometimes used on the frontend too, like mentioned table headings.

**Reply by brandonkelly** [reply]
Craft 4.5.6 is out with a slight improvement inspired by this: now any static row headings will get statically translated (9f4bf6e59f3f8cd7a1c791c60d9959bed4c84c1b).

(Column headings already got translated.)

**Comment by foxtrotcz** [comment]
So according to this issue https://github.com/craftcms/cms/issues/13763 it looks like row headings in CP are always statically translated according to CP language, not selected site language.

So I also propose there should be some way for row headings to be translated according to selected site language, not CP language. It would make sense in cases when row headings are treated as content and rendered by twig in frontend. So row headings language in backend should match the frontend.

I tried to make it work myself and its easily achievable by simple edit in https://github.com/craftcms/cms/blob/develop/src/fields/Table.php
Just define `$requestedSiteLanguage = \craft\helpers\Cp::requestedSite()->language;`
And use it as target language in `Craft::t();`

So basically replace this block:
```        
        // Apply static translations
        foreach ($defaults as &$row) {
            foreach ($this->columns as $colId => $col) {
                if ($col['type'] === 'heading' && isset($row[$colId])) {
                    $row[$colId] = Craft::t('site', $row[$colId]);
                }
            }
        }
```


By this block:

```        
        // Apply static translations
        $requestedSiteLanguage = \craft\helpers\Cp::requestedSite()->language;
        foreach ($defaults as &$row) {
            foreach ($this->columns as $colId => $col) {
                if ($col['type'] === 'heading' && isset($row[$colId])) {
                    $row[$colId] = Craft::t('site', $row[$colId],[],$requestedSiteLanguage);
                }
            }
        }
```

**Comment by foxtrotcz** [comment]
Still hope this will be improved someday :)

**Comment by foxtrotcz** [comment]
I tried via patching.
https://github.com/cweagans/composer-patches
https://github.com/symplify/vendor-patches

Using module would require to maintain the whole _normalizeValueInternal function which is unreasonable for single line change.
I came up with this. Not really best but it works.
In CP it uses Cp::requestedSite so I see translation based on site i currently edit and on public website it just gets null and uses default.

```
--- ../src/fields/Table.php
+++ ../src/fields/Table.php
@@ -510,7 +510,7 @@
         foreach ($defaults as &$row) {
             foreach ($this->columns as $colId => $col) {
                 if ($col['type'] === 'heading' && isset($row[$colId])) {
-                    $row[$colId] = Craft::t('site', $row[$colId]);
+                    $row[$colId] = Craft::t('site', $row[$colId], [], Cp::requestedSite()?->language);
                 }
             }
         }
```

**Comment by brandonkelly** [comment, marked as answer]
> So according to this issue #13763 it looks like row headings in CP are always statically translated according to CP language, not selected site language.

This is now fixed for the next release: e967d09baabf824898218d4f270a4b9d50ef5ec9

**Reply by foxtrotcz** [reply]
Thanks for this quick fix. I tried your change and it works. Now I can keep my CP language in English (which I prefer) and also see the Row Headings in the language of selected site I edit, so I can see what will be actually shown on the public website.

**Reply by brandonkelly** [reply]
Craft 5.9.23 is out with that change.
